### PR TITLE
[feat] `cast call --trace` (& code reuse in `cast run`)

### PR DESCRIPTION
**Pull request by nhtyy**
https://github.com/foundry-rs/foundry/issues/5461

Forks remote rpc, executes a call and prints a trace

edit: this PR also touches on `cast run` just to dedup some code we copy

## Motivation

I come up needing something like this quite often, using a forge test on the fly rn with calldata

### CODE REVIEW

**Comment by nhtyy** [comment]
hi, so this PR is fully implemented as a extension to `cast call`, although, im actually not sure if should instead become its own command like `cast trace`.

i only say this because now, `cast call` has options like `--debug` or `--verbose` that can only be used in trace mode, but they still appear at the top level `--help` call...

Making `--trace` a subcommand would alter the current semantics of how create calls are handled so im thinking this maybe should have its own command.

not sure.

**Review by mattsse — CHANGES_REQUESTED** [review]
great feature,

I think having it as flag in cast call is the best approach

I wonder if we can unify a lot of stuff with cast run

but perhaps not really possible since cast run behaves slightly differently

wdyt @evalir

only glanced over the changes, looks promising

I'd like to move the `TracingExecutor` and the standalone functions to a separate file though

**Review by Evalir — CHANGES_REQUESTED** [review]
looking good! some comments

**Comment by nhtyy** [comment]
hi @Evalir i addressed your requests in the more recent commits, as well as moved the executor to `evm::trace`, and moved things like `handle_trace`, `print_trace`, `run_debugger` to `cli::cmd::utils`. Just want to confirm those are good spots..

As well the all items mentioned can easily by re used in `cast run`! Should i do that here or should i make another pr?

File: cli/src/cmd/cast/call.rs [hunk 1]
```diff
@@ -31,13 +43,39 @@ pub struct CallArgs {
...
+    /// Can only be used with "--trace"
+    /// Labels to apply to the traces.
+    ///
+    /// Format: `address:label`
+    #[clap(long, requires = "trace")]
+    labels: Vec<String>,
+
+    /// Can only be used with "--trace"
+    #[clap(long)]
+    evm_version: Option<EvmVersion>,
```
Comment (Evalir) [inline cli/src/cmd/cast/call.rs:71]: actually i think this is OK to use without trace i believe—so you can call into contracts using shanghai or another incompatible evm version

File: cli/src/cmd/cast/call.rs [hunk 2]
```diff
@@ -73,54 +112,202 @@ pub enum CallSubcommands {
...
+                        }
+                    };
+
+                    handle_traces(trace, &config, chain, labels, verbose, debug).await?;
+
+                    return Ok(());
                 }
 
-                builder.set_data(data);
+                // fill the builder after the comditional so we dont move values
```
Comment (Evalir) [inline cli/src/cmd/cast/call.rs:205]:
```suggestion
                // fill the builder after the conditional so we dont move values
```

File: cli/src/cmd/utils.rs [hunk 3]
```diff
@@ -299,3 +307,114 @@ pub fn read_constructor_args_file(constructor_args_path: PathBuf) -> eyre::Resul
     };
     Ok(args)
 }
+
+/// A slimmed down return from the executor
```
Comment (Evalir) [inline cli/src/cmd/utils.rs:311]:
```suggestion
/// A slimmed down return from the executor used for returning minimal trace + gas metering info
```

File: cli/src/cmd/utils.rs [hunk 4]
```diff
@@ -299,3 +307,114 @@ pub fn read_constructor_args_file(constructor_args_path: PathBuf) -> eyre::Resul
...
+pub struct TraceResult {
+    pub success: bool,
+    pub traces: Traces,
+    pub debug: DebugArena,
+    pub gas_used: u64,
+}
+
+/// labels the traces, conditonally prints them or opens the debugger
```
Comment (Evalir) [inline cli/src/cmd/utils.rs:319]:
```suggestion
/// labels the traces, conditionally prints them or opens the debugger
```

File: cli/src/cmd/utils.rs [hunk 5]
```diff
@@ -299,3 +307,114 @@ pub fn read_constructor_args_file(constructor_args_path: PathBuf) -> eyre::Resul
...
+    Ok(())
+}
+
+pub async fn print_traces(
+    result: &mut TraceResult,
+    decoder: CallTraceDecoder,
+    verbose: bool,
+) -> eyre::Result<()> {
+    if result.traces.is_empty() {
+        eyre::bail!("Unexpected error: No traces. Please report this as a bug: https://github.com/foundry-rs/foundry/issues/new?assignees=&labels=T-bug&template=BUG-FORM.yml");
```
Comment (Evalir) [inline cli/src/cmd/utils.rs:370]: It's aight to error like this, but I think we can also just panic with `expect` and the hook will catch it and add the "this is a bug. please report" part :D

File: evm/src/trace/mod.rs [hunk 6]
```diff
@@ -25,8 +36,63 @@ pub mod identifier;
 
 mod decoder;
 pub mod node;
-mod utils;
+pub mod utils;
+
+/// a default executor with tracing enabled
```
Comment (Evalir) [inline evm/src/trace/mod.rs:41]:
```suggestion
/// A default executor with tracing enabled
```

**Review by Evalir — CHANGES_REQUESTED** [review]
lgtm bar few doc comments! can't push to your branch as it's protected (master)

i think we can do the other stuff in a followup pr if you're up for it 🙏

File: cli/src/cmd/utils.rs (hunk 5 above)
Comment (nhtyy) [reply cli/src/cmd/utils.rs:370]: i changed it to a panic!

File: cli/src/cmd/cast/call.rs (hunk 1 above)
Comment (nhtyy) [reply cli/src/cmd/cast/call.rs:71]: cast call just does an rpc call so im not sure of any place to set a version

Comment (nhtyy) [reply cli/src/cmd/cast/call.rs:71]: well while not using `--trace`

Comment (Evalir) [reply cli/src/cmd/cast/call.rs:71]: ah, fair enough! feel free to ignore then

**Comment by nhtyy** [comment]
@Evalir

just finished this, currently its a pr on my repo, wasnt sure where to point it

 https://github.com/nhtyy/foundry/pull/1

**Comment by Evalir** [comment]
@nhtyy nice! should we fix the failing tests on this pr and open that one here?

**Comment by nhtyy** [comment]
@Evalir 

https://github.com/foundry-rs/foundry/blob/1b994f70e81d666c77db97ea9ae0dc7d6b90aaf4/cli/src/cmd/cast/call.rs#L233-L247

should this `bail` on all variants of `EvmError`? I think no, i just copied it from the `cast run` but I think that it should handle the error case explicity like the snipped below, since `run` is only for published tx i guess this behavior makes sense? Although maybe it should be changed there to if my understanding is correct

https://github.com/foundry-rs/foundry/blob/1b994f70e81d666c77db97ea9ae0dc7d6b90aaf4/cli/src/cmd/cast/call.rs#L167-L197

but kk sounds good, ill fix the above and the fmt errors tonight,

**Comment by nhtyy** [comment]
@Evalir hmm sorry im not sure how to open both of them at the same time on this repo

**Comment by nhtyy** [comment]
maybe we should just merge them all here? diff shouldnt be to bad, sent you an invite to my repo as well

**Comment by Evalir** [comment]
sure! let's merge them all here then, and update the pr description accordingly

**Comment by nhtyy** [comment]
re-requesting review since there may have been some merge conflicts blocking

File: cli/src/cmd/cast/call.rs [hunk 7]
```diff
@@ -31,13 +40,39 @@ pub struct CallArgs {
...
+    verbose: bool,
+
+    /// Can only be used with "--trace"
+    /// Labels to apply to the traces.
+    ///
+    /// Format: `address:label`
+    #[clap(long, requires = "trace")]
+    labels: Vec<String>,
+
+    /// Can only be used with "--trace"
```
Comment (Evalir) [inline cli/src/cmd/cast/call.rs:66]:
```suggestion
    /// Can only be used with "--trace"
    ///
    /// The EVM Version to use.
```

**Review by Evalir — APPROVED** [review]
lgtm! thank you! just needs fmt, but can't push to your branch since this is on master. let's fix ci then we can send it

**Comment by meetmangukiya** [comment]
was missing this feature a lot! great stuff 🚀

File: evm/src/trace/mod.rs [hunk 8]
```diff
@@ -25,7 +36,64 @@ pub mod identifier;
...
+    fn deref(&self) -> &Self::Target {
+        &self.executor
+    }
+}
+
+impl DerefMut for TracingExecutor {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.executor
+    }
+}
```
Comment (Evalir) [inline evm/src/trace/mod.rs:96]: ah sorry, i believe we forgot this—can we move this to a separate file so it's not on mod? sorry for the late ask!

**Pull request merged**